src/main/java/com/github/packageurl/PackageURL.java

```diff
@@ -441,24 +442,49 @@ private String percentEncode(final String input) {
     }
 
     private static String uriEncode(String source, Charset charset) {
-        if (source == null || source.length() == 0) {
+        if (source == null || source.isEmpty()) {
             return source;
         }
 
-        StringBuilder builder = new StringBuilder();
-        for (byte b : source.getBytes(charset)) {
+        byte[] bytes = source.getBytes(charset);
+        int length = bytes.length;
+        int pos = indexOfFirstUnreservedChar(bytes);
+
+        if (pos == -1) {
+            return source;
+        }
+
+        StringBuilder builder = new StringBuilder(source.substring(0, pos));
```

Comment:
This optimization may not be more efficient than the previous code. `source.substring(0, pos)` allocates a new copy of the substring, which is immediately discarded by `new StringBuilder(…)`, which copies the contents of the string into a `StringBuilder` with enough capacity for exactly 16 additional characters beyond the end of the string (weird, but that's what the documentation says). If the encoded form is longer than the builder will need to reallocate. Maybe because it's uncommon for PURL components to be very long, the 16 characters works out.

Reallocation can be completely avoided by looping through once to determine the number of characters and again to populate the characters one at a time.

Reply:
It's "funny" that there's no constructor that takes a string and a capacity.

I removed this code and set the capacity to number of bytes * 3. But, I am not sure if it's worth creating a large buffer since in most cases, it won't be encoded at all. I am also not sure it's worth calculating the exact number vs. just setting an upper limit.

I didn't intend to do premature optimization, but the reason that one existing in the first place is to determine whether it even needs to be encoded so that it can return the original string (most common case, I think).